### ISSUE TITLE
`missing_docs` lint does not take into account the implementing type being `#[doc(hidden)]`

### ISSUE THREAD

**Issue by Nemo157** [rust-lang/rust#85526]
I tried [this code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=fbce6f3fc24a28d741173d074e9d5e3b):

```rust
#![warn(missing_docs)]

//! Crate docs

#[doc(hidden)]
pub struct Foo;

impl Foo {
    pub fn bar() {}
}
```

I expected to see this happen: no warnings, `Foo` and its method are not publicly documented.

Instead, this happened:

```
warning: missing documentation for an associated function
 --> src/lib.rs:9:5
  |
9 |     pub fn bar() {}
  |     ^^^^^^^^^^^^
  |
note: the lint level is defined here
 --> src/lib.rs:1:9
  |
1 | #![warn(missing_docs)]
  |         ^^^^^^^^^^^^

warning: 1 warning emitted
```

**Comment by pierwill** [comment]
Looks like the `tcx` here 

https://github.com/rust-lang/rust/blob/45c00481f3adbda69bd6f4e5816c74e2e2017f98/compiler/rustc_lint/src/builtin.rs#L528-L535

could give us the information needed to determine

1) if the item is an associated function definition, and 
2) if the type on which it is implemented itself has the `#[doc(hidden)]` attribute.

**Comment by pierwill** [comment]
[This method](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.associated_items) looks helpful.
